### ISSUE TITLE
Incorrect relative URL and missing fingerprint for bootstrap icons font

### ISSUE THREAD

**Issue by bjowes** [ligershark/WebOptimizer#325]
Our project uses Bootstrap, which includes icon fonts imported through CSS. After a recent update of the WebOptimizer library, the reference to the fonts is incorrect so they are not loaded by the site.

Path to bootstrap CSS: `/lib/bootstrap/dist/css/bootstrap.css`
Path to font file: `/lib/bootstrap/dist/fonts/glyphicons-halflings-regular.woff2`
Path to bundle where bootstrap CSS is included: `/css/bundle.css`

The import line in bootstrap CSS for the font file: `url("../fonts/glyphicons-halflings-regular.woff2")`

WebOptimizer 3.0.420 renders this in the bundle (correct relative path and fingerprint):
`url("../lib/bootstrap/dist/fonts/glyphicons-halflings-regular.woff2?v=93DUFX3R5ukJzB3V5fxFMdNdyXY")`

WebOptimizer 3.0.422 (and up) renders this in the bundle (incorrect relative path, no fingerprint):
`url("../../lib/bootstrap/dist/fonts/glyphicons-halflings-regular.woff2")`

The only change between those versions was #320, so the new behavior was introduced there. I spent some time looking through the changes but I'm not familiar enough with the code base to identify the exact root cause.

We can roll back to 3.0.420 for now but I expect that others will bump into this issue too.

**Comment by daesk** [comment]
I just noticed the same issue. 
Will downgrade, but it would be nice if it would be fixed.

**Comment by nvirth** [comment]
Same here. Worked with `3.0.396`, not working after upgrade to `3.0.426`

**Comment by nabukhas** [comment]
Same happened to me, I had to downgrade to version 3.0.420 until it getting fixed in future version.
the issue exist in versions: 3.0.422 and 3.0.426

**Comment by emperador-ming** [comment]
Same here. We had to downgrade to version 3.0.420. The issue still persists in 3.0.422 & 3.0.426.

**Comment by Elimi81** [comment]
Tried to fix it, but my knowledge of the library is too limited I'm afraid. I found out that the commit bf0933112a3ede2077bc9df4fe4a6f757b1daa7e breaks something when it changes the RelativePathAdjuster. In Nuget if you use version before 3.0.422 everything still works.

**Comment by ZephyrZiggurat** [comment]
@alexreinert do you have any ideas on this? the RelativePathAdjuster is adding additional `..` to css url()'s. it may just be when the app is deployed to a subfolder because it works fine locally (no <base> present anywhere).

**Comment by emperador-ming** [comment]
The issue still persists in 3.0.433.

**Comment by VahidN** [comment]
Only Version="3.0.413" works for me.

**Comment by Elimi81** [comment]
@jzabroski Hi! Thank you for all your recent contributions to this library :)

About this issue, i'm not sure what might be the cause of this issue but i created a small example that highlights the issue and how the new version (> 3.0.420) breaks the relative urls in css:

[Example project](https://github.com/Elimi81/WeboptimizerExample)

If you run the project, it should start in /test-url since it's in launchSettings.json and the icons don't work:

![Image](https://github.com/user-attachments/assets/6bb05121-1598-4756-aaf0-a4c9e1becc6b)

You can then "fix" this bu downgrading the WebOptimizer to version 3.0.420 OR by running the application without the /test-part in launchSettings.json. Either of these changes will make the icons work again:

![Image](https://github.com/user-attachments/assets/e36e372d-89d8-4f2b-9e1a-893b991e4b8d)

So any insight on the issue is appreciated. Thanks!

**Comment by jzabroski** [comment]
Thanks. I see you are using a PublishProfile, which is a feature I've never used.

I also see this line:
https://github.com/Elimi81/WeboptimizerExample/blob/0152058d8ddb64aaae617cb6067abc427b3ce153/UIProject/Program.cs#L7

Can you explain to me how `_content` and its subdirectories is making its way into your final deployment? I'm just not familiar with this approach. I don't know how to debug a pubxml. Maybe it's an irrelevant detail, but I am assuming you thought it was important to have two projects instead of one for me to reproduce this issue? Is this the smallest possible example?

**Comment by Elimi81** [comment]
> Thanks. I see you are using a PublishProfile, which is a feature I've never used.
> 
> I also see this line: https://github.com/Elimi81/WeboptimizerExample/blob/0152058d8ddb64aaae617cb6067abc427b3ce153/UIProject/Program.cs#L7
> 
> Can you explain to me how `_content` and its subdirectories is making its way into your final deployment? I'm just not familiar with this approach. I don't know how to debug a pubxml. Maybe it's an irrelevant detail, but I am assuming you thought it was important to have two projects instead of one for me to reproduce this issue? Is this the smallest possible example?

The publishprofile is just there because I tried the same application on our server and the publishprofile contains information on how and where to publish the application on IIS. The issue can be seen even when running this on localhost, so the publishprofile does not affect this. Infact publishprofile does not affect anything on runtime, it's just to publish the application wherever you want to publish it.

The _content-path is build in functionality on Razor Library and [documented behavior](
https://learn.microsoft.com/en-us/aspnet/core/blazor/components/class-libraries?view=aspnetcore-9.0&tabs=visual-studio#create-an-rcl-with-static-assets-in-the-wwwroot-folder)

Also I don't think that is the cause either since even the icon hosted directly in the application does not work.

**Comment by Elimi81** [comment]
I have removed the Razor Library from the example to make it as small as possible.

**Comment by jzabroski** [comment]
Thanks, I will look at it today. To be honest I have been mostly out of web development since 2013, so the Razor behavior you cited was not familiar to me. I do a few pages every now and then to chip in when people are stretched thin, but I actually focus primarily on database stuff. I'm back at it now because I'm vibe coding upgrading ASP.NET MVC5 to ASP.NET Core on .NET 8.0. Lots of fun vibe coding.

**Comment by Elimi81** [comment]
> 2. In [Bugfix for Razor Class Library assets #287](https://github.com/ligershark/WebOptimizer/pull/287) , [@gumbarros](https://github.com/gumbarros) introduced a new `GetAssetFileProvider()` extension method. This change is half-done-in-the-oven in that there are still 3 references to the old `GetAssetProvider()` extension method. It seems obvious to me here that the call to .Last() will skip the _content folder. In this sense, [@Elimi81](https://github.com/Elimi81) , your initial repro is helpful as it demonstrates the flaws in GetFileProvider() - it will never work for Razor Class Libraries since its skipping the content root. It would have to use the ContentRootFileProvider to work with Razor Class Libraries.

I'm not sure what you mean by "it" when you said it will never work? Css-files have always been working in Razor Class Libraries with WebOptimizer and we have been using them for a very long time. After the commit I highlighter earlier the thing that is not working are the relative paths inside those css-files (after minification).

I was now looking into this again and this is what I found out.

Prerequisite of these is you are running the application in url like localhost/myapp etc.

| Version | Asset added to bundle | Works? |
|--------|--------|--------|
| < 3.0.422 | /css/site.css | YES |
| < 3.0.422 | css/site.css | NO |
| >= 3.0.422 | /css/site.css | NO |
| >= 3.0.422 | css/site.css | YES |

So what should i take from this? Should I change the way we add our assets to the bundle going forward or is this a bug in WebOptimizer?

**Comment by jzabroski** [comment]
A bit more progress here. It turns out the original motivation for the FileProvider feature was #51 c.f. this comment by @madskristensen **(emphasis mine)** :

> This is currently not supported. When registering assets on the pipeline your only choice in terms for modifying the IFileProvider is to use the ContentRootProvider using the UseContentRoot() method.
> 
> For this to work, we need to be able to pass in a custom file provider when we register the WebOptimizer service.
> 
> **In your case where you have multiple file providers (one for each module), we should combine them all into a CompositeFileProvider.**
> -- @madskristensen , https://github.com/ligershark/WebOptimizer/issues/51#issuecomment-340486188

Since ASP.NET Core 1.0 was released on June 27, 2016, and Mads comment was on October 30, 2017, it may not have been a fully developed use case then.

There appear to be 3 scenarios:
1. Simplcommerce - distributed across many modules, each with their own dependencies, feeding into a main Website Application project
2. Razor Class Libraries - which put resources in the _content (Content Root)
3. SImple Web Applications - which put resources in the wwwroot (Web Root)

Since the ContentRoot and Web Root are directly exposed over DI, and Static Files "Root" can be virtually resolved by registering a CompositeFileProvider e.g. like so:

```csharp
app.UseStaticFiles(new StaticFileOptions() 
{
    FileProvider = new CompositeFileProvider(
        new PhysicalFileProvider(Folder1),
        new PhysicalFileProvider(Folder2)
    )
}); 
```
The remaining thing is to register a file provider with WebOptimizer, as would be necessary if you wanted to use something outside the Content Root and Web Root. That being said, the CompositeFileProvider that is registered by default in ASP.NET Core for the Content Root should be sufficient for most users, unless you override it with a different file provider.

tldr I'm getting clarity on what the clean fix is here. For some people it won't be a drop-in fix where they take an update and it just works. Code changes may be required. Will think through how to avoid breaking change by default.

**Comment by gburghardt** [comment]
I pulled the solution down and imported WebOptimizer.Core into another existing Git repo as a sub module. I could add a direct project reference and debug this line-by-line. I can confirm that adding `app.UseBasePath("...")` in Program.cs triggers the bug. Removing that line causes the bug to go away.

It looks like RelativePathAdjuster.cs is where things go awry somewhere between [line 61](https://github.com/ligershark/WebOptimizer/blob/8651c1bca1e585b24338890beef7643cb0249d4f/src/WebOptimizer.Core/Processors/RelativePathAdjuster.cs#L61) and [line 89](https://github.com/ligershark/WebOptimizer/blob/8651c1bca1e585b24338890beef7643cb0249d4f/src/WebOptimizer.Core/Processors/RelativePathAdjuster.cs#L89). It is probably in the [MakeRelative](https://github.com/ligershark/WebOptimizer/blob/8651c1bca1e585b24338890beef7643cb0249d4f/src/WebOptimizer.Core/Processors/RelativePathAdjuster.cs#L94) method or maybe in `UrlPathUtils.MakeAbsolute`.

I am calling `app.UseBasePath(string)` before calling `app.UseRouting()`.

I'm going to have a good honest sitdown with my debugger this afternoon to see if I can narrow it down.

**Comment by emperador-ming** [comment]
@gburghardt, did you finally get that chance to sit down and have an honest talk with your debugger? It really felt like you were almost there!

**Comment by jzabroski** [comment]
Me having an honest talk with my debugger on a GitHub Issue I said I would investigate further...

[<img width="498" height="358" alt="Image" src="https://github.com/user-attachments/assets/30bba18d-f79c-409d-98f0-c92ed09ffd9c" />](https://tenor.com/view/liliths-throne-gif-3770370188439759195)

In all seriousness, maybe I will point Claude at it. I have not been tokenmaxxing my subscription of late.
